server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -485,4 +544,163 @@ private static String signatureMessage(final Class<? extends Plugin> clazz) {
     public <T> List<T> filterPlugins(Class<T> type) {
         return plugins.stream().filter(x -> type.isAssignableFrom(x.instance().getClass())).map(p -> ((T) p.instance())).toList();
     }
+
+    static final LayerAndLoader createPluginModuleLayer(PluginBundle bundle, ClassLoader parentLoader, List<ModuleLayer> parentLayers) {
+        assert bundle.plugin.getModuleName().isPresent();
+        return createModuleLayer(
+            bundle.plugin.getClassname(),
+            bundle.plugin.getModuleName().get(),
+            urlsToPaths(bundle.urls),
+            parentLoader,
+            parentLayers
+        );
+    }
+
+    static final LayerAndLoader createSpiModuleLayer(
+        String moduleName,
+        Set<URL> urls,
+        ClassLoader parentLoader,
+        List<ModuleLayer> parentLayers
+    ) {
+        // assert bundle.plugin.getModuleName().isPresent();
+        return createModuleLayer(
+            null,  // no entry point
+            moduleName + ".spi",  // by convention ?
```

Comment:
The spi jar is separate, it should have module-info as well right? I guess we don't pull this right now...we can probably leave this as a followup, but we should probably actually read it (and require it is there if the main plugin is modularized)